### PR TITLE
dns authenticator, fixes #200

### LINKED ISSUE #200: implement dns authenticator
authentication (proof of domain control) for ACME will be possible via dns txt records.

dns updates (of dynamic zones) can be done via the nsupdate cmdline tool or any other rfc2136 compatible client (e.g. the dnspython library has code for that), so that one does not have to automatically modify configuration or data files of misc. nameserver implementations.

i think this authenticator might be nice for various reasons:
- there are lots of different webserver softwares out there and we can't support them all like the few most popular ones [and even for the popular ones, it might take some time until done], so an alternative way of authentication is needed
- if one does not want a certificate for https usage, but something else, using an external webserver just to prove control is a bit weird
- does not depend on a specific nameserver as it uses rfc2136 dynamic updates (nameserver just needs to support that)
- with a fake "nsupdate" tool that just prints stdin and waits for a key press, manual dns updates for zones without dynamic update capability are possible
- nameserver does not need to be on same machine as no config file access or restarting is required
- does not require root for authentication

### PR DESCRIPTION
**Pull request by ThomasWaldmann**
the authenticator code is covered by unit test (I looked at the apache tests as a base for this), but it is not really integrated into letsencrypt yet. also, the LE server does not support dns auth yet.

the low level code (that invokes nsupdate cli tool) has no unit tests as it depends on that tool being installed and a dynamically updateable dns server + key for it being available. but I did manual tests, it works. :)

please review, there are some TODOs / unclear things in the code.

### CODE REVIEW

File: letsencrypt/client/ddns/ddns.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
+"""generic DDNS authenticator (RFC 2136)"""
+
```
Comment (kuba) [inline letsencrypt/client/ddns/ddns.py]: unnecessary whitespace

File: letsencrypt/client/ddns/ddns.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import zope.interface
+
+from letsencrypt.client import challenge_util
+from letsencrypt.client import interfaces
+
+
+class DDNS(object):
+    """DDNS authenticator.
+
```
Comment (kuba) [inline letsencrypt/client/ddns/ddns.py]: unnecessary whitespace

File: letsencrypt/client/tests/ddns/ddns_test.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        self.assertEqual(mock_nsupdate.call_count, 2)
+        calls = mock_nsupdate.call_args_list
+        expected_call_list = [
+            ("add", "encryption-example.demo",
+             "17817c66b60ce2e4012dfad92657527a", ),
+            ("add", "letsencrypt.demo",
+             "27817c66b60ce2e4012dfad92657527a", ),
+        ]
+
+        for i in range(len(expected_call_list)):
```
Comment (kuba) [inline letsencrypt/client/tests/ddns/ddns_test.py]: pls use `xrange` instead

File: letsencrypt/client/tests/ddns/ddns_test.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+    @mock.patch("letsencrypt.client.ddns.ddns.nsupdate")
+    def test_perform2(self, mock_nsupdate):
+        responses = self.ddns.perform(self.challs)
+
+        self.assertEqual(mock_nsupdate.call_count, 2)
+        calls = mock_nsupdate.call_args_list
+        expected_call_list = [
+            ("add", "encryption-example.demo",
+             "17817c66b60ce2e4012dfad92657527a", ),
```
Comment (kuba) [inline letsencrypt/client/tests/ddns/ddns_test.py]: pls remove the trailing comma (`, )`); here and below

File: letsencrypt/client/tests/ddns/ddns_test.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        self.assertEqual(mock_nsupdate.call_count, 2)
+        calls = mock_nsupdate.call_args_list
+        expected_call_list = [
+            ("del", "encryption-example.demo",
+             "17817c66b60ce2e4012dfad92657527a", ),
+            ("del", "letsencrypt.demo",
+             "27817c66b60ce2e4012dfad92657527a", ),
+        ]
+
+        for i in range(len(expected_call_list)):
```
Comment (kuba) [inline letsencrypt/client/tests/ddns/ddns_test.py]: `xrange`

File: letsencrypt/client/tests/ddns/ddns_test.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    def test_cleanup2(self, mock_nsupdate):
+        self.ddns.cleanup(self.challs)
+
+        self.assertEqual(mock_nsupdate.call_count, 2)
+        calls = mock_nsupdate.call_args_list
+        expected_call_list = [
+            ("del", "encryption-example.demo",
+             "17817c66b60ce2e4012dfad92657527a", ),
+            ("del", "letsencrypt.demo",
+             "27817c66b60ce2e4012dfad92657527a", ),
```
Comment (kuba) [inline letsencrypt/client/tests/ddns/ddns_test.py]: trailing commas in tuples

**Comment by kuba** [comment]
Nice, thanks. Minor things to fix, otherwise LGTM. Please also create API autodoc.

File: letsencrypt/client/ddns/ddns.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+            if isinstance(chall, challenge_util.DnsChall):
+                nsupdate("del", chall.domain, chall.token)
+
+
+# TODO: make this user-configurable.
+NSUPDATE_CMD = "nsupdate -k le-nsupdate.key"
+
+
+def nsupdate(action, domain, token):
+    """Invoke the nsupdate commandline tool to send a single DNS update
```
Comment (kuba) [inline letsencrypt/client/ddns/ddns.py]: Isn't there a Python library for that?

**Comment by kuba** [comment]
`dnsutils` should probably be listed in the dependencies if we choose to stick to the `nsupdate` binary

File: letsencrypt/client/ddns/ddns.py (hunk 7 above)
Comment (kuba) [inline letsencrypt/client/ddns/ddns.py]: How about http://www.dnspython.org/? Example `nsupdate`-like tool: https://gist.github.com/pklaus/4619865 that shows how to use `dns.update` module.

File: letsencrypt/client/tests/ddns/ddns_test.py (hunk 3 above)
Comment (ThomasWaldmann) [inline letsencrypt/client/tests/ddns/ddns_test.py]: right. but that was just c&p from apache-related code, iirc. :)

Comment (kuba) [inline letsencrypt/client/tests/ddns/ddns_test.py]: fixed in #205 ;)

File: letsencrypt/client/ddns/ddns.py (hunk 7 above)
Comment (ThomasWaldmann) [inline letsencrypt/client/ddns/ddns.py]: we can do that, I use dnspython in my other project nsupdate.info.
so if you prefer python code, even if it adds a rather big library, i can do that.

it likely needs some more code on our side also, to parse the keys from a bind-like key file and simulate some magic, like the nsupdate cli tool does.

File: letsencrypt/client/tests/ddns/ddns_test.py (hunk 6 above)
Comment (ThomasWaldmann) [inline letsencrypt/client/tests/ddns/ddns_test.py]: I like them (also in lists, dicts, sets).
0 element thing: [] () {}
n+1 element thing: add element and comma

I find this more natural than the other way. Also, if you have each element on a separate line, it makes the diff smaller if you add a new element (as you just have to add that new line, not also a comma at the end of the previous line).

File: letsencrypt/client/ddns/ddns.py (hunk 2 above)
Comment (ThomasWaldmann) [inline letsencrypt/client/ddns/ddns.py]: i like to separate stdlib, dependencies, own code imports. and also keep them in that order.

File: letsencrypt/client/tests/ddns/ddns_test.py (hunk 3 above)
Comment (jdkasten) [inline letsencrypt/client/tests/ddns/ddns_test.py]: Xrange should be used based on the implementation (I just read up on it), my only concern would be that xrange isn't in python3.  Xrange was relabeled as range, but I think we have bigger problems to worry about when/if we try to transition.

Comment (ThomasWaldmann) [inline letsencrypt/client/tests/ddns/ddns_test.py]: there will be some compatibility library that chooses correctly xrange (py2) or range (py3), so this is a minor problem. supporting 2.6 (and py3 <3.3 also) are bigger problems.

File: letsencrypt/client/ddns/ddns.py (hunk 2 above)
Comment (jdkasten) [inline letsencrypt/client/ddns/ddns.py]: I believe he meant the whitespace under the module docstring

**Comment by ThomasWaldmann** [comment]
i'll add dnsutils requirement to (new) docs after merge into master.

File: letsencrypt/client/tests/ddns/ddns_test.py (hunk 6 above)
Comment (kuba) [inline letsencrypt/client/tests/ddns/ddns_test.py]: I also like trailing comas :)... provided you can justify it by the "diff smaller" argument. Note that here trailing comma is followed by closing bracket, which would require you to change 2 lines anyway. Please observe the rules (here for `[]`, but they apply to different collection types and their respective braces):
```python
[1, 2, 3, 4, 5]  # yes
[1, 2, 3, 4, 5, ]  # no

[
    1,
    2,
    3,
]  # yes
[
    1,
    2,
    3
]  # no
```
To stay consistent with the rest of the code base we should either stick to the rules above or refactor everywhere. Of course it's a matter of personal preference, but IMHO comma followed by bracket just looks weird and less readable..

File: letsencrypt/client/ddns/ddns.py (hunk 7 above)
Comment (kuba) [inline letsencrypt/client/ddns/ddns.py]: I'd strongly prefer to use library (`dnspython`), can you do that please? :)

File: letsencrypt/client/ddns/ddns.py (hunk 2 above)
Comment (kuba) [inline letsencrypt/client/ddns/ddns.py]: nope, I meant class docstring (`DDNS.__doc__`)

**Comment by kuba** [comment]
@TW new docs is pulled in, but ofc `dnsutils` is not necessary after you update the PR to use `dnspython`

File: letsencrypt/client/ddns/ddns.py (hunk 2 above)
Comment (jdkasten) [inline letsencrypt/client/ddns/ddns.py]: :) I didn't open the whole file.  Yes the DDNS docstring should just be one line. 

(removing the first line whitespace does keep it consistent with the rest of the codebase though too)  Thanks @ThomasWaldmann!

File: letsencrypt/client/tests/ddns/ddns_test.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        # XXX unclear: what is key?
+        key = None
+        self.challs = []
+        self.challs.append(challenge_util.DnsChall(
+            "encryption-example.demo", "17817c66b60ce2e4012dfad92657527a", key))
+        self.challs.append(challenge_util.DnsChall(
+            "letsencrypt.demo", "27817c66b60ce2e4012dfad92657527a", key))
+
+    def tearDown(self):
+        pass
```
Comment (kuba) [inline letsencrypt/client/tests/ddns/ddns_test.py]: `teadDown` empty -> remove?

File: letsencrypt/client/tests/ddns/ddns_test.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+    def test_pref(self):
+        result = self.ddns.get_chall_pref("example.org")
+        self.assertEqual(result, ["dns"])
+
+    def test_perform0(self):
+        resp = self.ddns.perform([])
+        self.assertTrue(resp == [])
```
Comment (kuba) [inline letsencrypt/client/tests/ddns/ddns_test.py]: `assertEqual(resp, [])`?

**Comment by ThomasWaldmann** [comment]
My suggestion is to keep nsupdate cli tool for now. when the server counterpart is implemented and stable, we can later switch to dnspython if it is worth the effort. Read the nsupdate man page about misc. sorts of magic / additional stuff that nsupdate cli tool does which we would partly or all have to re-implement when using dnspython.

Also, my idea was to have the NSUPDATE_CMD configurable, so an admin can modify params (what kind of auth is used, where to get key, keyfile format, ...), use some wrapper or even another tool (that would process same input).

File: letsencrypt/client/ddns/ddns.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+from letsencrypt.client import challenge_util
+from letsencrypt.client import interfaces
+
+
+# TODO: make practical use of this, nothing uses DDNS yet
+class DDNS(object):
+    """DDNS authenticator.
+
+    """
```
Comment (jdkasten) [inline letsencrypt/client/ddns/ddns.py]: Still extra whitespace on 14-15.  I can take care of this on the eventual merge if you don't end up taking another pass.

**Comment by jdkasten** [comment]
Hi @ThomasWaldmann.  Thanks for writing this up.

If you don't want to continue working on this right now I suggest we move it into a dns_authenticator branch.  There won't really be any conflicts as the only shared file it touches is setup.py.

This code isn't hooked up for available use yet, and the interface isn't yet practical.  We also have some permission integration issues to work through. My natural inclination is also towards the dnspython route though I haven't taken the time to investigate the issue and the magic provided by the binary :)

I think moving it to a separate branch will allow for continued development from more contributors.

Thoughts?

**Comment by kuba** [comment]
Okay, I managed played a bit with the code It'd be awesome if you could somehow clarify my questions below and update documentation, so that devs like me, who have never used `nsupdate` before, will be able to easily maintain code in the future. :)

`man nsupdate` doesn't say anything about return codes; how do you make sure that operation actually succeeds and DNS is indeed updated? Does the code take care of this somehow? Is it worth to read `nsupdate add ...` stdout to check it? Could you please add an appropriate documentation?

Is `nsupdate` CLI consistent across different operating systems? Across different packages providing it (output for Debian sid below)?

```
$ apt-file search '/usr/bin/nsupdate'
dnsutils: /usr/bin/nsupdate
knot-dnsutils: /usr/bin/nsupdate
```

I guess that the above problems would be solved by using `dnspython`, or at least bring as closer to universal and testable solution.

Also, where does `le-nsupdate.key` come from (how to generate it, links to relevant docs)? How do you test for its presence? Once I somehow get the key, where do I put it, given paths are relative? This should be made configurable and documented as well.

Speaking of tests, [coverage for this module is only 62%](https://travis-ci.org/letsencrypt/lets-encrypt-preview/jobs/48742861#L1429) and it has to be 100% before merging into `master`. As you noted, however, we might want to wait before server is capable of issuing DNS challenges.

**Comment by ThomasWaldmann** [comment]
as you both seem to be very keen on dnspython, let me just state some issues / limitations you will run into when going that route. you may have noticed that i personally do use dnspython in another project rather than nsupdate, but that is under different circumstances (for that other project, there is usually only 1 relatively capable server admin [not a lot of more-or-less capable admins/site owners] using this functionality and there is no need for interactivity).

support
----------
if you use dnspython, people will come to you and say "it doesn't work", just because they didn't get their nameserver setup right (there will be a lot of authentication issues, it will depend on the nameserver and authentication used) and you will have to deal with it as it will just fail in letsencrypt's code.

if you used nsupdate cli tool, this is very nicely handled by some yet-to-write docs that tell "first get nsupdate working on the commandline [this is totally NOT letsencrypt's project's business], then configure that commandline for letsencrypt". No issues inside letsencrypt that way.

hacks / mods
------------------
People will be able to easily modify how nsupdate works, by trying out different cmdline options (for authentication, for debug mode, ...). They also can put their key whereever they want, etc.
They even can exchange nsupdate binary for a self-written shell script that just outputs stdin and then waits for a keypress. That way you can manually deal with updating a nameserver via a web form or whatever.
You can't do that this flexible (for a user) with a builtin nsupdate clone using dnspython.

**Comment by ThomasWaldmann** [comment]
@jdkasten "This code isn't hooked up for available use yet, and the interface isn't yet practical."

Not hooked up: yes. Some pointers how it should be hooked up?

"We also have some permission integration issues to work through."

Not sure what you mean.

**Comment by ThomasWaldmann** [comment]
@kuba return codes: i tried to update a host where I do not have permissions to do so. nsupdate returned a non-zero return code then, as it is usually done on unix. for a successful update, it did return 0. And you see in the code that precisely that is checked and there is an exception if it is non-zero (same as check_call/check_output would use). I didn't see the need to look into stdout/stderr yet, this is just logged for debug log level.

consistency across packages / OSes: no idea, I just tried on ubuntu. As the rc == 0 is pretty standard on POSIX, i would assume it works there. Not that sure about windows, but I guess most of the other code won't work there anyway.

le-nsupdate.key must be made by the user, see "man nsupdate". I personally used the same key snippet as is in the bind configuration. The user will have to configure bind anyway (or whatever he uses), so that part is not really in-scope of letsencrypt. The key can be put anywhere and the user may give any absolute or relative path that works, that's why NSUPDATE_CMD should be configurable.

Test coverage: it will never be possible to be 100% there except if you provide a stable dynamically updateable nameserver for test purposes. In other parts of the code I saw you did not test the parts where test code would reach outside over the internet, but you rather used mocking, so I did it the same.

**Comment by jdkasten** [comment]
@ThomasWaldmann If we are going to push it into master we should try to do it right.  There is no need to rush, because no one is blocking on it.

We need to adapt the CLI to allow for the DNS authenticator.  It should probably require a specific command because it doesn't quite fit the determine_authenticator scheme in main.py.  I envision the people using it will be specifically targeting it.  Those who do not ask for it most certainly do not want to mess with it.

Something along these lines
https://cloud.githubusercontent.com/assets/1581662/5482522/dde0a4aa-8617-11e4-830e-814fcc473392.jpg

I can live without it being hooked up right now, but I think the big problem is the current nsupdate interface doesn't allow anyone to use it even if they did hook it up.  You can't hardcode in keys and expect the user to go through the code and change them.  I know it is marked as a TODO, but why not fix it?  Without it, it isn't usable (and this whole project is about usability).  We should probably abstract away from the binary as much as possible to make it more general.

DNS is a touchy subject and I have seen concerns online about the client implementation within ACME.  I want to be confident that everything works/looks appropriate before we merge into master.  For instance, right now if the program crashes in between putting up the challenge and cleaning up, I believe the DNS records remain.

Permission integration issues:
If the goal is to remove the root requirement with this, it would also be nice to think through everything before we merge this into master.  The main transition can be in a separate branch... but we should be doing things like dropping to the minimum set of privileges before we call out to the binary.

Testing:
Mocking does not mean you can't have 100% coverage.  For instance, to test lines 72-85 you can mock subprocess.Popen specifically instead of the whole nsupdate function.

**Comment by ThomasWaldmann** [comment]
The permissions issue is rather about every other code *except* the one in this pull request. You can run this code as unprivileged user and call nsupdate. So I suggest this is dealt with not here but in the ticket that is open about requesting root late (and only when appropriate). If you don't need root you should not first require it and then drop privileges.

cli: there is a yet unanswered question in the ticket #136. maybe the code should not do anything by default, not even try to configure your (maybe non-existing) apache, but the user should request that (via cli param or ncurses).

about user configuration: i have seen you have a client/CONFIG.py, but that is as well part of the code as everything else (and if letsencrypt is a system package, installed to read-only system locations). so, how shall it work?

about crashes: yes, that is true, there is no builtin rollback. but it's not a big issue. there's no harm caused by a txt record nor it is a problem to use nsupdate manually to delete that record in such hopefully rare cases.

mocking Popen: ah, nice idea. did you do that somewhere else where you call Popen?

File: letsencrypt/client/tests/ddns/ddns_test.py [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+class DDNSPerformTest(unittest.TestCase):
+    """Test the DNS challenge (for dynamic DNS zones)."""
+
+    def setUp(self):
+        super(DDNSPerformTest, self).setUp()
+
+        from letsencrypt.client.ddns import ddns
+        self.ddns = ddns.DDNS()
+
+        # XXX unclear: what is key?
```
Comment (ThomasWaldmann) [inline letsencrypt/client/tests/ddns/ddns_test.py]: btw, this is a really missing thing yet and still unclear.

**Comment by ThomasWaldmann** [comment]
I moved the nsupdate commandline to CONFIG.py. As I said: that doesn't solve the issue you pointed out, but at least you have all these issues at one place now. :)

File: letsencrypt/client/ddns/ddns.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
+"""generic DDNS authenticator (RFC 2136)"""
```
Comment (kuba) [inline letsencrypt/client/ddns/ddns.py]: `"""generic` -> `"""Generic`

**Comment by ThomasWaldmann** [comment]
fixed.

**Comment by jdkasten** [comment]
Sorry for the delay on this... these responses take quite a bit of time to write up.
Yes, we do mock subprocess.Popen a few times in the project.  Here is one of example
https://github.com/letsencrypt/lets-encrypt-preview/blob/c5db07cf0a5be2d3d7d86271bca50aca3ca36fd8/letsencrypt/client/tests/apache/util.py#L67

The permissions issue is based on the fact that the initial scope of the project wasn't to support such a feature.  Those goals have changed but there is still a bit of work.  I know that I will eventually take care to make sure that we minimize the attack surface, but right now, master isn't in that condition and this pull strictly increases the attack surface and assumes changes are already made in master.   I haven't actually looked at nsupdate and all of its functions, but normally network binaries are dangerous.  I am sure this will come up and I will spend more time answering questions and responding to criticism than it would take to actually take to implement the feature correctly.  (This is more of a kick at me to get everything else in place... I just don't want to have to deal with the temporary fallout)

I didn't envision this going in as a config parameter, no user changes are/were or should be needed in the CONFIG file in order to get a certificate under typical use.  We also were assuming root so it wasn't that big of an issue.  Luckily, @kuba is remedying that issue and making a user supplied config file.  I would still like to be able to specify everything on the command line.  I imagine users would be more likely just to run it and assume that they don't have to change a config file.  If you do want to rely on a CONFIG parameter for now, I believe the default CONFIG parameter should be blank with a comment above about how to set it.  That way when people run it without setting it...you can fail gracefully and tell them what to do.

@ThomasWaldmann Is it just the key parameter that needs to be specified?

The initial idea as specified in the CLI #136 was to simply determine the best course of action if nothing was specified.  This was stubbed in the determine_authenticator, determine_installer functions.  If an Apache web server isn't installed, the code should throw a NoInstallation error (although I haven't actually made it around to truly hardening the ApacheConfigurator code so right now it throws a NoInstallation error if it can't find the Apache config files.)  The idea was that the code would determine what was installed and if it could make an intelligent decision... and go with it.  For example, if we know Apache is installed and it is listening on port 80 and 443... assume that it is the correct server.  If we know that nothing is listening on port 443 use the standalone authenticator #63.  Until this logic can be fleshed-out, I think the default will be to add a menu of the different authenticators that the system can find.  So when things get ironed out here... it would be Apache, Stand-alone, DNS... or stand-alone, DNS if Apache isn't installed.  Though most people who are editing DNS are probably (and for good reason) not on the same machine as their web server.  That is why I initially said it was a different use case.  (Forewarning: I am currently cleaning the display system).  This also isn't a block to getting it merged.

Regarding the DNS TXT records with crashes, I agree that it isn't a major issue.  However, being responsible for maintaining this code into the foreseeable future, I can see someone complaining about littering their DNS records.  Perhaps if we add better error handling code to avoid control-c from exiting before running cleanup we can mitigate this issue.  We also have prexisting code that I can add that would eliminate this.  This isn't a block from merging.

**Comment by ThomasWaldmann** [comment]
just let the user specify the full nsupdate commandline needed. that is simpler and more flexible than splitting this into pieces.

**Comment by jdkasten** [comment]
Specifying the whole command on the commandline seems extremely risky to me....

"We will just take whatever you specify here and *run* it."

If anyone tries to write a program that uses letsencrypt as a tool we just allowed them to get pwned.  I do not believe that is an option.

**Comment by kuba** [comment]
In regards to `CONFIG`: #220

**Comment by ThomasWaldmann** [comment]
@jdkasten maybe you understood me wrong. I didn't say to specify the whole nsupdate commandline *on the (letsencrypt) commandline*. I said let the user specify the whole nsupdate commandline (and I meant putting that into the configuration file, which I did, see this diff).

If i understood #220 correctly, this creates an equivalence between configuration and commandline though, right? So THAT might be a problem, but that is not in scope of THIS PR.

**Comment by ThomasWaldmann** [comment]
increased test coverage with mocking of Popen

**Comment by jdkasten** [comment]
It doesn't matter if it is on the command line or is in a user supplied config file.  It is and always will be data that is specific to the user.  Every user will have a different path to their key and every user must supply the field.  In all sense of the word, it is data.  This PR treats it directly as code...

Separation of data and code is standard secure coding practice 101.  This is a security application... we *must* follow the basic principles.  I absolutely cannot accept a PR that violates the principle so blatantly.  (I would be thrown out of my program for such an infraction.)

This is not a hard change to make...

**Comment by ThomasWaldmann** [comment]
@jdkasten I'ld suggest you discuss all topics relating to user supplied config in the pull request #220 which is exactly about that. It has nothing to do with the code supplied by this pull request.

also, this PR does not have a cli interface, so there is no issue with it.

that said, you are calling quite some other binaries, even as root, and they are configurable so I suggest you first care for these and don't push all of letsencrypts problem's into the discussion of this PR, esp. if they have nothing to do with the code in this PR.

**Comment by kuba** [comment]
I'm afraid that @jdkasten is absolutely right... There is a difference between specifying path to a binary and specifying full command line! Please update your code correspondingly, i.e. introduce `NSUPDATE_PATH` and `NSUPDATE_KEY`.

Also, please update to match the style guidelines provided.

Anyway, why do we keep on discussing this? Using `dnspython` would eliminate this attack vector. @ThomasWaldmann, have you already started working on this?

Also, this has nothing to do with #220. User-supplied config will always be input by users, no matter whether it's CLI argument or a file lying on the disk (also note that config file path would probably have to be specified through CLI itself...).

**Comment by ThomasWaldmann** [comment]
@kuba show me file and line number in my PR that allows an attacker to specify the command line.

and no, i am for sure not gonna work on using dnspython as long as there is not even a server counterpart.

File: letsencrypt/client/CONFIG.py [hunk 13]
```diff
@@ -116,3 +116,6 @@
 
 APACHE2 = "/etc/init.d/apache2"
 """Command used for reload and restart."""
+
+# TODO: add dnsutils to required/suggested packages
+NSUPDATE_CMD = "nsupdate -k le-nsupdate.key"  # see "man nsupdate"
```
Comment (kuba) [inline letsencrypt/client/CONFIG.py]: @ThomasWaldmann this is the offending line.

Note, that although `master` does not yet have #220 merged, `CONFIG` was always meant to be editable by users. Even if not (we might argue how hard it is to edit Python code), then this creates unnecessary vulnerability I will have to take care of when merging #220, and since it's your code, fixing it now would be awesome (you get the attribution as well!). Thanks :)

**Comment by kuba** [comment]
FYI I'm planning to work on the DNS challenge support in Boulder soon.

**Comment by ThomasWaldmann** [comment]
Can we pleeeeeeease stay in-scope of *this* PR?

The line you pointed to is in program code and it is currently either not editable by users/attackers (when the code is safely installed to system locations as it should be if you care for security). Or *any* code is editable by users/attackers, then discussing about security is pointless anyway.

File: letsencrypt/client/ddns/ddns.py [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+             in a non-zero return code
+
+    """
+    cmd = CONFIG.NSUPDATE_CMD
+    logging.debug("nsupdate cmd: %s", cmd)
+    stdin = "update %s _acme-challenge.%s. 60 TXT %s\nsend\n" % (
+        action, domain, token)
+    logging.debug("nsupdate stdin: %s", stdin)
+    process = subprocess.Popen(
+        cmd.split(),
```
Comment (kuba) [inline letsencrypt/client/ddns/ddns.py]: Okay, you're right, @ThomasWaldmann, I should have pointed at this line instead... Think for a second on what "modular" code is. `CONFIG` is meant to be editable by users and any application code, here `ddns.nsupdate`, should sanitize the input properly. You fail to filter the input, thus allowing virtually any command to be passed in. If anyone happens to refactor `CONFIG` to be actually editable (just happens to be #220), we are screwed, just because you didn't follow the convention! :(

Comment (ThomasWaldmann) [inline letsencrypt/client/ddns/ddns.py]: See #221 and EOF here about that stuff.

**Comment by jdkasten** [comment]
It isn't about the binaries, it is about the mixing of the two....user input and code.  I know it is likely that 99.9% of the people using the tool will be actual users on the command line.  They won't be trying to give bad input as they already have arbitrary execution.  I understand that... but I also believe it should clearly be separate and I believe people will point to it as sloppy.  (It certainly doesn't inspire confidence.  I also plan on scrubbing all of the input... TODO in main.py) People will be looking for a reason for the project to fail (people I have already personally dealt with) and I will spend more time defending such an action than it would take to simply fix it.  This isn't a problem... but as I suggested in the very first comment.... at this point in time it makes more sense to simply pull it into a separate branch.  (Which I have now done)  I think this is very close to complete, but when I fully bring it in (and take over some responsibility for it) I don't want to have to worry about it.

There are still a few more test cases to write and I haven't actually tested out the code myself.  (It is difficult without the server challenges).

If a separate branch is fine for right now then I say we close this PR.
